File: backpack/backpack.cpp

using namespace std;

#include<iostream>
#include<vector>
#include <algorithm>



class Solution {
public:
    /**
     DP solution
     to see the dp when dp is the status of rest value;
     */
    int backPack(int m, vector<int> &A) {
		vector<int> val(m + 1, 0);
        
        for(int i = 0; i < (m + 1); i++){
            val[i] = 0;
        }

        for(int i = 0; i < A.size(); i++){
            for(int j = m; j > 0; j--){
                if(j >= A[i]){
                    val[j] = max(val[j], val[j-A[i]]+A[i]);
                }
            }
        }
		return val[m];
    }
};

int main(int argc, char const *argv[])
{
    vector<int> a = { 20,15,15,15,15,15,15,15,15,15,15,15,15,15,15,7,2,15,15,15,15 };
    cout<< Solution().backPack(10, a) << endl;
	system("pause");
    return 0;
}

File: backpack/backpack.py


class Solution:
    """
    Error: time limit
    this is a typical pathon run-time-limit problem
    same code as cpp, problem solved;
    """
    def backPack(self, m, A):
        val = [0 for i in range(m+1)]

        for i in range(0, len(A)):
            for j in range(m, 0, -1):
                if j >= A[i]:
                    val[j] = max(val[j], val[j - A[i]]+A[i])
                    # print('node[%d] rest[%d] val[%d]' %(i-1, j,val[j]))
        return val[m]

if __name__ == "__main__": 
    print(Solution().backPack(10, [3,4,8,5]))
